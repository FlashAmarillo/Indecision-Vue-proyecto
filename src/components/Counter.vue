<template>
  <h2>{{ customTitle }}</h2>
  <p data-testid="counter"> {{ counter }}</p>
  <p> {{ counter }} <sup>2</sup> = {{ squareCounter }}</p>
  <div class="botones-container">
      <button @click="increase">+1</button>
      <button @click="decrease">-1</button>
  </div>

  <footer>
      <p>Diomar Villarroel 2022</p>
  </footer>
</template>

<script>
export default {
    //properties/propiedades o props (definimos sus tipos de datos)
    props: {
        title: String,
        start: {
            type: Number,
            required: true,
            default: 100,
            validator( value ) {
                return value >= 0
            }
        }
    },

    name: 'Patito',
    data() {
        return {
            counter: this.start
        }
    }, 
    methods: {
        getSquareValue() {
            return this.counter * this.counter;
        },
        //incremento al clickear +1
        increase() {
            console.log('Aumentado');
            this.counter++;
        },
        //decremento al clickear -1
        decrease() {
            console.log('Disminuyendo')
            this.counter--;
        }
    }, 
    computed: { //Las propiedades computadas se guardan en el cache haciendolas m√°s eficientes
        squareCounter() {
            return this.counter * this.counter
        },
        customTitle() {
            console.log(this.title)
            return this.title === undefined ? 'Counters' : this.title;
        }
    }
}
</script>

<style>

button {
    background-color: #64bb87;
    border-radius: 5px;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    font-size: 1.3rem;
    margin: 0 5px;
    padding: 10px 30px;
    transition: 0.3s ease-in-out;
}

button:hover {
    background-color: #5aa67b;
    transition: 0.3s ease-in-out;
}

</style>