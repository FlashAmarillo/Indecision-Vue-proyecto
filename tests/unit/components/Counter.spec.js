import { shallowMount } from "@vue/test-utils";
import Counter from "@/components/Counter";

describe('Counter component Vue', () => {
    // definimos el wrapper de forma global, me va a crear una instancia del componente Counter montado y renderizado en el DOM 
    let wrapper;

    // antes de cada prueba, reinicia el wrapper
    beforeEach( () => {
        wrapper = shallowMount( Counter );
    })

    test('H2 debe tener el valor por defecto "Counter"', () => {
        
        expect( wrapper.find('h2').exists()).toBeTruthy();

        const h2Velue = wrapper.find('h2').text();

        expect(h2Velue).toBe('Counters');

    })

    test('El valor por defecto debe ser 100 en el parrafo P', () => {
        
        // traemos todos los parrafos
        const parrafo = wrapper.find('[data-testid="counter"]').text();

        // expect segundo p = 100
        expect(parrafo).toBe("100");

    })

    test('debe incrementar y decrementar el valor del contador al hacer click en los boton', async () => {

        //tomar la referencia de los botones
        const [increaseBtn, decreaseBtn] = wrapper.findAll('button');

        // le montamos un click al boton con el metodo trigger 3 veces
        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');
        // hacemos click 2 veces segudias para reducir
        await decreaseBtn.trigger('click');
        await decreaseBtn.trigger('click');

        // evaluamos si aumento el state
        const value = wrapper.find('[data-testid="counter"]').text();
        expect(value).toBe('101');
    })

    test('debe de establecer el valor por defecto', () => { 
        
        // obtenemos el valor de los props
        const { start } = wrapper.props();

        const value = wrapper.find('[data-testid="counter"]').text();

        // Evaluamos que el valor sea igual al prop por defecto
        expect(Number(value)).toBe(start);
    })

    test('Debe de mostrar la prop de title', () => { 

        const title = "Hola Mundo Diomar!!!";

        // le pasamos props a nuestro componente Counter
        const wrapper = shallowMount(Counter, {
            props: {
                title
            }
        });

        expect( wrapper.find('h2').text() ).toBe(title);


        console.log(wrapper.html());
     })
})