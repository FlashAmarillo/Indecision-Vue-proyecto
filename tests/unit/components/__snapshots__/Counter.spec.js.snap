// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component Vue Debe hacer match con el snapshot 1`] = `
<h2>Counters</h2>
<p>100 <sup>2</sup> = 10000</p>
<div><button>+1</button><button>-1</button></div>
<footer>
  <p>Diomar Villarroel 2022</p>
</footer>
`;
